### PR TITLE
⚠️ Generate v1alpha4 types

### LINKED ISSUE #3428: Generate v1alpha4 types
Tracking issue to generate Cluster API types for:

- Core types
- CABPK
- KCP
- CAPD

/kind feature
/area api
/milestone v0.4.0

### PR DESCRIPTION
**Pull request by srm09**
**What this PR does / why we need it**:
Generates types for v1alpha4
- [ ] Core types
- [ ] Core/experimental types
- [ ] Bootstrap types
- [ ] Control plane types
- [ ] Docker infra types

Remove v1alpha2 types

**Which issue(s) this PR fixes**:
Fixes #3428

### CODE REVIEW

File: test/infrastructure/docker/main.go [hunk 1]
```diff
@@ -28,12 +28,12 @@ import (
...
-	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
-	clusterv1exp "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	clusterv1exp "sigs.k8s.io/cluster-api/exp/api/v1alpha4"
 	"sigs.k8s.io/cluster-api/feature"
-	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha3"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha4"
 	"sigs.k8s.io/cluster-api/test/infrastructure/docker/controllers"
-	infrav1exp "sigs.k8s.io/cluster-api/test/infrastructure/docker/exp/api/v1alpha3"
+	infrav1exp "sigs.k8s.io/cluster-api/test/infrastructure/docker/exp/api/v1alpha4"
```
Comment (srm09) [inline test/infrastructure/docker/main.go:36]: IMO this might need both alpha3 and new alpha4 refs to be added to the scheme.

**Comment by srm09** [comment]
@vincepri what's the deal with the `go-apidiff`? How do I go about fixing it?

**Comment by srm09** [comment]
@vincepri Running into a [similar issue](https://github.com/kubernetes-sigs/controller-runtime/issues/810) with auto generated conversion functions.
https://github.com/kubernetes-sigs/cluster-api/pull/3749/commits/8f93d8c6413f55d846ca0d603bb04eb0f514458e#diff-3b87981b2f78c1c597ebd6a3f2326fffR162

**Comment by vincepri** [comment]
> @vincepri what's the deal with the go-apidiff? How do I go about fixing it?

That's normal that it's failing, it detects breaking changes from the previous commit, it's also an optional job

> @vincepri Running into a similar issue with auto generated conversion functions in controlplane/kubeadm/api/v1alpha3/zz_generated.conversion.go.

Let's sync up today, @ncdc updated the related issue yesterday with more details. I can't think of any immediate workaround, unless we can overwrite the conversion functions that are referring cross package types

**Comment by ncdc** [comment]
@srm09 @vincepri this is what I think we need to do to generate & use the right conversions:
- Each API package needs to use a unique build tag
- We need to invoke conversion-gen separately, once per API package
- When we need to use conversion functions from packages outside the one we're generating, we have to specify those packages as `--extra-peer-dirs` using the **fully qualified package name**

All of these changes can be made in the Makefile.

Here's a start to the conversion-gen calls that we'll need
- 1 for ./api/v1alpha3, build-tag = something like `ignore_autogenerated_core_v1alpha3`
- 1 for ./bootstrap/kubeadm/api/v1alpha3, extra-peer-dirs = sigs.k8s.io/cluster-api/api/v1alpha3, build-tag ~ `ignore_autogenerated_kubeadm_bootstrap_v1alpha3`
- etc

Basically start with the foundation and then build up, adding extra-peer-dirs as needed.

**Comment by srm09** [comment]
@ncdc The changes that you suggested above to add `build` tags and `extra-peer-dirs` to the makefile targets for code generation. I added the bare minimums for this PR, and I can create a separate issue and PR entirely dedicated to all the changes you mentioned above.

File: test/infrastructure/docker/main.go (hunk 1 above)
Comment (srm09) [reply test/infrastructure/docker/main.go]: Added the v1alpha4 schemes too

**Comment by srm09** [comment]
@ncdc I added the following things to Makefile for generation of conversion functions:
- **core**: now has a `--build-tag=ignore_autogenerated_core_v1alpha3`
- **bootstrap-kubeadm**: has `--build-tag=ignore_autogenerated_kubeadm_bootstrap_v1alpha3 --extra-peer-dirs=sigs.k8s.io/cluster-api/api/v1alpha3` which causes the generated functions to use functions from `api/v1alpha3 or api/v1alpha4` instead of unsafe conversions
- **kubeadm-control-plane**: has `--build-tag=ignore_autogenerated_kubeadm_controlplane_v1alpha3 --extra-peer-dirs=sigs.k8s.io/cluster-api/api/v1alpha3,sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha3` which produces the similar kind of file.
- **infra/docker**: has `--build-tag=ignore_autogenerated_infra_v1alpha3 --extra-peer-dirs=sigs.k8s.io/cluster-api/api/v1alpha3`

### Caveat/Learning
A puzzling thing that I came across was, when adding the `--build-tag` flag the first time, the generated conversion file would **omit** the generation of functions of the type `Convert_v1alpha3_KubeadmControlPlane_To_v1alpha4_KubeadmControlPlane`, just generate the functions of type `autoConvert_v1alpha3_KubeadmControlPlane_To_v1alpha4_KubeadmControlPlane`. Another run would go ahead and generate both these function types. I tried to look at the conversion-gen source code, but could not get anywhere with that.

File: test/infrastructure/docker/Makefile [hunk 2]
```diff
@@ -113,6 +113,9 @@ generate-go: $(CONTROLLER_GEN) $(CONVERSION_GEN) ## Runs Go related generate tar
 		paths=./$(EXP_DIR)/api/...
 	$(CONVERSION_GEN) \
 		--input-dirs=./api/v1alpha3 \
+		--input-dirs=./$(EXP_DIR)/api/v1alpha3 \
+		--build-tag=ignore_autogenerated_infra_v1alpha3 \
```
Comment (vincepri) [inline test/infrastructure/docker/Makefile:117]:
```suggestion
		--build-tag=ignore_autogenerated_capd_v1alpha3 \
```

Comment (vincepri) [reply test/infrastructure/docker/Makefile:117]: Maybe?

Comment (srm09) [reply test/infrastructure/docker/Makefile:117]: I kept it as `infra` since the imports in the Go files were prefixed with the same.

Comment (srm09) [reply test/infrastructure/docker/Makefile]: Changed it to `capd`

**Comment by ncdc** [comment]
@srm09 re your caveat about the Convert_ functions taking 2 conversion-gen invocations to be generated, I think that's a problem we'll need to resolve. I can dig into it later if needed.

**Comment by srm09** [comment]
The second invocation runs on top of the changes from the first invocation, and generates the missing conversion functions. I can spend some time looking into it as well, but am not sure what is the underlying requirement? Maybe adding a flag to generate/omit the generation of functions of type `Convert_` might make the behavior more predictable.

File: Makefile [hunk 3]
```diff
@@ -613,3 +627,6 @@ diagrams: ## Build proposal diagrams
 serve-book: ## Build and serve the book with live-reloading enabled
 	$(MAKE) -C docs/book serve
 
+.PHONY: remove-generated-conversion-files
+remove-generated-conversion-files: ## Remove files generated by conversion-gen from the mentioned dirs
+	(IFS=','; for i in $(SRC_DIRS); do find $$i -type f -name 'zz_generated.conversion*' -exec rm -f {} \;; done)
```
Comment (vincepri) [inline Makefile:632]:
```suggestion
.PHONY: remove-generated-conversion-files
clean-generated-conversions: ## Remove files generated by conversion-gen from the mentioned dirs
	(IFS=','; for i in $(SRC_DIRS); do find $$i -type f -name 'zz_generated.conversion*' -exec rm -f {} \;; done)
```

**Comment by vincepri** [comment]
@srm09 Let's squash the last few commits, this should be good to go

@ncdc what do you think?

**Comment by srm09** [comment]
@vincepri Currently, the commits are arranged according to the generation of types for each of the sub-project, core, CAPBK.. 
Am thinking of:
- Keeping the generation of each sub project in a separate commit
- Merging the removal of v1a2, update to v1a4 import types and Makefile changes into a single commit.

That should result in about 6 commits, is that fine?

**Comment by srm09** [comment]
@ncdc Is this waiting on any more changes to be merged to the default branch?

**Review by vincepri — APPROVED** [review]
/approve
/assign @ncdc 
for final lgtm

**Comment by ncdc** [comment]
@srm09 let me give it another 👀 before merging.

**Comment by srm09** [comment]
> New changes are detected. LGTM label has been removed.

Had to fix tests after rebase

File: bootstrap/kubeadm/main.go [hunk 4]
```diff
@@ -53,9 +55,11 @@ func init() {
 
 	_ = clientgoscheme.AddToScheme(scheme)
 	_ = clusterv1alpha3.AddToScheme(scheme)
```
Comment (ncdc) [inline bootstrap/kubeadm/main.go:57]: I'm not sure we need to add both 3 and 4 to the scheme. We can just do 4, right?

File: bootstrap/kubeadm/main.go [hunk 5]
```diff
@@ -53,9 +55,11 @@ func init() {
...
+	_ = clusterv1alpha4.AddToScheme(scheme)
 	_ = expv1alpha3.AddToScheme(scheme)
```
Comment (ncdc) [inline bootstrap/kubeadm/main.go:59]: I'm not sure we need to add both 3 and 4 to the scheme. We can just do 4, right?

File: bootstrap/kubeadm/main.go [hunk 6]
```diff
@@ -53,9 +55,11 @@ func init() {
...
-	_ = kubeadmbootstrapv1alpha2.AddToScheme(scheme)
+	_ = expv1alpha4.AddToScheme(scheme)
 	_ = kubeadmbootstrapv1alpha3.AddToScheme(scheme)
```
Comment (ncdc) [inline bootstrap/kubeadm/main.go:61]: I'm not sure we need to add both 3 and 4 to the scheme. We can just do 4, right?

File: controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+
+	cabpkv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/errors"
+)
+
+const (
+	KubeadmControlPlaneFinalizer = "kubeadm.controlplane.cluster.x-k8s.io"
+
+	// DEPRECATED: This label has been deprecated and it's not in use anymore.
```
Comment (ncdc) [inline controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go:31]: Might as well remove this now?

File: controlplane/kubeadm/main.go [hunk 8]
```diff
@@ -51,8 +54,11 @@ func init() {
 
 	_ = clientgoscheme.AddToScheme(scheme)
 	_ = clusterv1alpha3.AddToScheme(scheme)
+	_ = clusterv1alpha4.AddToScheme(scheme)
```
Comment (ncdc) [inline controlplane/kubeadm/main.go:57]: Same comment as in another main.go - I think we only need the alpha4 APIs added to the scheme

File: main.go [hunk 9]
```diff
@@ -75,10 +77,12 @@ func init() {
 	klog.InitFlags(nil)
 
 	_ = clientgoscheme.AddToScheme(scheme)
-	_ = clusterv1alpha2.AddToScheme(scheme)
 	_ = clusterv1alpha3.AddToScheme(scheme)
+	_ = clusterv1alpha4.AddToScheme(scheme)
```
Comment (ncdc) [inline main.go:81]: Same as before - just v1alpha4?

File: test/infrastructure/docker/main.go [hunk 10]
```diff
@@ -30,12 +30,16 @@ import (
 	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
 	"k8s.io/klog"
 	"k8s.io/klog/klogr"
-	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
-	clusterv1exp "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	clusterv1alpha3 "sigs.k8s.io/cluster-api/api/v1alpha3"
```
Comment (ncdc) [inline test/infrastructure/docker/main.go:33]: I'd undo these changes, go back to the clusterv1, infrav1, etc aliases, just change the imports to v1alpha4

File: test/infrastructure/docker/main.go [hunk 11]
```diff
@@ -59,10 +63,14 @@ func init() {
 	klog.InitFlags(nil)
 
 	_ = scheme.AddToScheme(myscheme)
-	_ = infrav1.AddToScheme(myscheme)
-	_ = infrav1exp.AddToScheme(myscheme)
-	_ = clusterv1.AddToScheme(myscheme)
-	_ = clusterv1exp.AddToScheme(myscheme)
+	_ = infrav1alpha3.AddToScheme(myscheme)
```
Comment (ncdc) [inline test/infrastructure/docker/main.go:66]: Should only need alpha4 added to scheme

**Review by ncdc — CHANGES_REQUESTED** [review]
@srm09 thanks for your hard work on this! A few questions:

- Do we need to update cmd/clusterctl/... in this PR (it still imports v1alpha3 and the CRDs it owns are still v1alpha3)?
- test/framework/clusterctl/... still has v1alpha3 imports?
- Some of the files in test/infrastructure/docker/{examples,templates} still use v1alpha3?
- controllers/cluster_controller.go imports sigs.k8s.io/cluster-api/exp/api/v1alpha3 - this probably needs to be change to v1alpha4 and let's update the import alias to expv1?
- The following still have sample/test data using v1alpha3 (may be ok, though?)
    - util/util_test.go
    - util/yaml/yaml_test.go
    - controlplane/kubeadm/internal/machinefilters/machine_filters_test.go
    - controlplane/kubeadm/controllers/helpers_test.go
    - exp/controllers/machinepool_controller_phases_test.go
    - exp/controllers/machinepool_controller_test.go
    - controllers/machine_controller_phases_test.go
    - controllers/mdutil/util_test.go
    - controllers/machine_controller_test.go

File: controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go (hunk 7 above)
Comment (srm09) [reply controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go:31]: Makes sense, is there a set convention that we follow to remove unused stuff from CAPI codebase? Or do we follow the upstream k8s conventions?

File: main.go (hunk 9 above)
Comment (srm09) [reply main.go:81]: The existing file had imports to both v1a2 and v1a3 packages, hence followed suite.

More of a generic question, what is the scheme used for? FWIU, this is used to register types in a particular version. Wouldn't then the v1a3 types be needed for upgrade purposes?

Comment (ncdc) [reply main.go:81]: The scheme allows us to encode/decode specific GroupVersionKinds (e.g. a v1alpha3 Cluster or a v1alpha4 Cluster). If all our controllers are only talking v1alpha4, there shouldn't be any need to add v1alpha3 to the scheme.

Comment (srm09) [reply main.go:81]: And since we registered conversions for v1a3 to v1a4 using the h-and-s model, the controllers would only ever use v1a4 types. Is that the correct assumption?

**Comment by srm09** [comment]
> Do we need to update cmd/clusterctl/... in this PR (it still imports v1alpha3 and the CRDs it owns are still v1alpha3)?

The issue did not list the creation of the `clusterctl` types, hence did not include those in this PR.

> test/framework/clusterctl/... still has v1alpha3 imports?

same as above

> The following still have sample/test data using v1alpha3 (may be ok, though?)
> util/util_test.go
> util/yaml/yaml_test.go
> controlplane/kubeadm/internal/machinefilters/machine_filters_test.go
> controlplane/kubeadm/controllers/helpers_test.go
> exp/controllers/machinepool_controller_phases_test.go
> exp/controllers/machinepool_controller_test.go
> controllers/machine_controller_phases_test.go
> controllers/mdutil/util_test.go
> controllers/machine_controller_test.go

The test data just points to the API version, not sure if it is really necessary to change it. But if we wanna get rid of the `v1alpha3` references and keep a single version number in tests around, I can update those.

**Comment by ncdc** [comment]
/lgtm

🚀 🎉 🎂
